note to expand context for empty alt vs no alt

scottaohara opened the pull request:
[Fixes #885](https://github.com/w3c/html/issues/885)

Adding a note to provide additional context to how screen readers will interpret images that have an alt set to the empty string, vs no alt attribute set.

including code examples to further explain how an empty alt could be problematic.

Would definitely appreciate any feedback on the prose here / if anyone has thoughts on if this needs to be a more succinct note.

Thank you

scottaohara committed:
note to expand context for empty alt vs no alt

Adding a note to provide additional context to how screen readers will interpret images that have an alt set to the empty string, vs no alt attribute set.

including code examples to further explain how an empty alt could be problematic.

scottaohara committed:
some rewording

chaals commented on sections/semantics-embedded-content.include:
```diff
@@ -3623,12 +3623,63 @@
-  altogether, under the assumption that the image is a key part of the
-  content.
+  As a last resort, implementors should set the <code>alt</code> attribute to the empty
+  string, under the assumption that the image is a purely decorative image that doesn't add any
```
As a last resort, implementors should leave out the alt: https://www.w3.org/TR/ATAG20/#gl_b23

chaals commented on sections/semantics-embedded-content.include:
```diff
@@ -3623,12 +3623,63 @@
+  <div class="note">
+    Setting an <{image}> element's <code>alt</code> attribute to the empty string will convey
+    to assistive technologies, such as screen readers, that the image in question provides no
+    substantial informative value, and should be considered decorative to the section of the 
```
and to content management / analysis systems

chaals commented on sections/semantics-embedded-content.include:
```diff
@@ -3623,12 +3623,63 @@
+    least indicate to a user that an image existed in the document.
+
+    For example, while the following image has no alternative text, it will be discoverable
+    by assistive technologies as an image, and its file name should be announced.
```
There is no particular mandate to announce the filename - but whatever information a user agent / AT chooses to provide as a repair attempt might include the filename. Or maybe not...

chaals commented on sections/semantics-embedded-content.include:
```diff
@@ -3623,12 +3623,63 @@
+    As the <code>alt</code> attribute is set to the empty string, in this example, assistive
+    technologies will likely not announced the existence of the image. Thus to users of such
+    software, the <{figcaption}> would be describing "nothing" as there would be no announcement
+    of a "figure" in this instance.
```
equally, content management tools may assume that the image is decorative, and therefore not attempt to find an appropriate alternative for it.

chaals committed:
add atag reference, tweak text

stevefaulkner commented on sections/semantics-embedded-content.include:
```diff
@@ -3623,12 +3623,67 @@
+    an image without an <{img/alt}> attribute is preferable to an <{img/alt}>
+    attribute set to the empty string.
+
+    If an image has no alaternative text, screen readers will generally announce there is an image,
```
typo - _alaternative_

scottaohara committed:
fix typo

LJWatson commented on sections/semantics-embedded-content.include:
```diff
@@ -3623,12 +3623,63 @@
+    Setting an <{image}> element's <code>alt</code> attribute to the empty string will convey
+    to assistive technologies, such as screen readers, that the image in question has no relative
+    or informative value to the section of the document in which it is found. Being that it serves
+    no informative purpose, a screen reader may choose to completely ignore the presence of the
```
Suggest changing this to "a screen reader will..." instead of "a screen reader may...", since all screen readers will completely ignore an image with an empty alt attribute.

LJWatson commented on sections/semantics-embedded-content.include:
```diff
@@ -3623,12 +3623,63 @@
+
+    However, in instances where an image is vital to understanding the contents of a document,
+    an image with an omitted <code>alt</code> attribute is preferable to an <code>alt</code>
+    attribute set to the empty string. In the former example, assistive technology would at
```
Minor nit, but suggest changing "would" to "will", to be consistent with the tense of the rest of the content.

LJWatson commented on sections/semantics-embedded-content.include:
```diff
@@ -3623,12 +3623,63 @@
+      <figure>
+        <img src="bar-graph.png">
+        <figcaption>
+          <p>
```
Does there need to be a ```<p>``` element inside the ```<figcaption>```?

LJWatson commented on sections/semantics-embedded-content.include:
```diff
@@ -3623,12 +3623,63 @@
+    the empty string:
+
+    <xmp highlight="html">
+      <figure>
```
It might be better to use a different code example for these two examples. When an image with no alt is inside a figure and there is a figcaption, the figcaption provides the accessible name for the image. An alternative may be to use a simple ```<img>``` example, but include the expected accessible name for the image in each case (so slience when alt="", and filename when the alt is missing).

scottaohara replied:
Thanks @LJWatson.

I'll take a stab at revising these examples this evening.

scottaohara commented:
Taking both @chaals and @LJWatson's edits & reviews into consideration, I've revised the prose further.  Thank you both for helping me along with this update.

scottaohara committed:
revised examples and prose

taking Chaal’s revised content, and the suggestions from Léonie, I’ve revised the note to be shorter than my original submission, while better integrating the examples with the prose, and reducing the prose further.

chaals commented on sections/semantics-embedded-content.include:
```diff
@@ -3623,27 +3623,58 @@
+    should be aware that an image without an <{img/alt}> attribute is preferable to an
+    <{img/alt}> attribute set to the empty string.
+
+    Taking the previous example and omitting the <code>alt</code> attribute, as such:
```
As such reads strange to me. I would find "as follows" or something like that more natural. But this is editorial, feel free to ignore.

chaals commented on sections/semantics-embedded-content.include:
```diff
@@ -3623,27 +3623,58 @@
+  <div class="note">
+    Setting an <{image}> element's <{img/alt}> attribute to the empty string means the image in
+    question provides no essential information. Assistive technology, such as screen readers,
+    will completely ignore the presence of such an image, as ignoring it will not stop the user
```
s/completely/typically/ ??

As far as I know that is the normal and rational behaviour, but I don't think it is the only possible behaviour. Allowing for other cases can reduce confusion when someone who has a strange setup - e.g. because they are not a habitual screen reader user - finds that something different occurs.

YMMV though - this is a question/suggestion rather than a formal request for change

chaals reviewed (approved):
A couple of comments that may be addressed by ignoring them.

please go ahead and decide then merge.

scottaohara committed:
wording updates
